src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -330,7 +330,10 @@ private void MessageWasProbablySent(CommandRequestMessage message)
             }
 
             var transaction = _session.CurrentTransaction;
-            if (transaction != null && transaction.State == CoreTransactionState.Starting)
+            if (transaction != null &&
+                transaction.State == CoreTransactionState.Starting &&
+                message.WasSent) // we should not change a transaction state if a client side error has been thrown
```

Comment:
You should probably test `message.WasSent` in the caller (even though it's two places) because if the message was not sent we don't want to call this method at all.

Reply:
>if the message was not sent we don't want to call this method at all.
but we do it for now. As I understand the method name(`probably`) tells us that there was an attempt to send the message, not that the message has been sent. If so, I guess, it's ok to have this validation inside this method. I consider it like we have attempted to send the message and now, we need to update different statuses based on the sending results

Reply:
I realize we do it now.

But if `message.WasSent` is false then we know 100% that the message was not sent, and it would be better not to call this method.

Reply:
Maybe not 100% now that I think of it.

If an exception is thrown when writing to the socket `WasSent` will be false even though we may (or may not) have actually sent it.

So OK.

Reply:
> Maybe not 100% now that I think of it.

Actually maybe yes. Unless the exception was thrown when encoding.

Looks like we set `WasSent` to `true` immediately after encoding the message but before writing it to the socket.

Reply:
So it still looks like if `message.WasSent == false` then there is no chance the message was sent to the server and we shouldn't call `MessageWasProbablySent` because it wasn't.

Reply:
done